templates/traffic-agent-rbac.yaml

```diff
@@ -0,0 +1,135 @@
+  labels:
+    app.kubernetes.io/name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+    app.kubernetes.io/part-of: {{ .Release.Name }}
+    helm.sh/chart: {{ include "ambassador.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    {{- if .Values.deploymentTool }}
+    app.kubernetes.io/managed-by: {{ .Values.deploymentTool }}
+    {{- else }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+    {{- end }}
+    product: aes
+---
+## Create a service-account-token for traffic-agent with a matching name.
+## Since the ambassador-injector will use this token name, it must be deterministic and not auto-generated.
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    kubernetes.io/service-account.name: traffic-agent
+type: kubernetes.io/service-account-token
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: Role
+metadata:
+  name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app.kubernetes.io/name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+    app.kubernetes.io/part-of: {{ .Release.Name }}
+    helm.sh/chart: {{ include "ambassador.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    {{- if .Values.deploymentTool }}
+    app.kubernetes.io/managed-by: {{ .Values.deploymentTool }}
+    {{- else }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+    {{- end }}
+    product: aes
+rules:
+  - apiGroups: [""]
+    resources: [ "namespaces", "services", "secrets" ]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: [ "getambassador.io" ]
+    resources: [ "*" ]
+    verbs: ["get", "list", "watch", "update"]
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app.kubernetes.io/name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+    app.kubernetes.io/part-of: {{ .Release.Name }}
+    helm.sh/chart: {{ include "ambassador.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    {{- if .Values.deploymentTool }}
+    app.kubernetes.io/managed-by: {{ .Values.deploymentTool }}
+    {{- else }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+    {{- end }}
+    product: aes
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ include "ambassador.rbacName" . }}
+subjects:
+  - name: {{ .Values.servicePreview.trafficAgent.serviceAccountName }}
+    namespace: {{ .Release.Namespace }}
+    kind: ServiceAccount
+{{- else }}
+## If we install Service Preview cluster-wide, this means we can't use the 'traffic-agent' ServiceAccount
+## as it does not exist in every namespace. We must instead grant new Roles to all ServiceAccounts (cluster-wide).
```

Comment:
I am actually quite concerned by this but I do not know a better way 🤔

Reply:
This is one of the rare cases where single-namespace is easier.

I honestly do not see a better way, that would allow for a quick & easy installation through `edgectl install` and make Service Preview accessible without constantly reconfiguring ServiceAccounts. 

It is definitely not intended for production use!

Reply:
If that is the case would we be able to have a separate RBAC configuration for service preview? Is this even entirely reliable if the user does not normally have a service account in their pod definition?

Is it completely unreasonable to have the injector be able to create a service account?

Reply:
> If that is the case would we be able to have a separate RBAC configuration for service preview?

It is here. We are creating distinct `traffic-manager` and `traffic-agent` ServiceAccounts with a subset of the roles.

> Is this even entirely reliable if the user does not normally have a service account in their pod definition?

Yes, because in any case a "default" ServiceAccount is injected in every pod. This default SA typically has no privileges.

> Is it completely unreasonable to have the injector be able to create a service account?

`MutatingWebhookConfiguration` such as the Injector here should have no side effect on Kubernetes resources. If the Injector itself was to create a ServiceAccount, it would lead us in a weird state where the K8s control loop would already have begun for the Pod resource lifecycle but has to reset this loop waiting for the other resources to be available. Plus, this would also mean granting SA creation, and RoleBinding permissions to the injector... :/

---

The "true" solution to the problem we are facing here would require a major architectural change to Service Preview. The injected sidecar should only be the Envoy "data-plane" and require no privileges. This sidecar would depend on an external long-lived "control-plane" which has read permissions to configure the said data-plane. This is the model used by service meshes.

Reply:
Okay then the change I would like to see is to separate the scope of the RBAC installations. i.e you should be able to install Ambassador with cluster rbac and service preview with namespaced